### PR TITLE
add capability for explicit train/test/validate paths

### LINKED ISSUE #34: Change the way datasets are imported
It would be better if the datasets was specified differently in the configure file. 
Now, we are using numbers to indicate which dataset to use. 
It would be much more convenient if the user specified himself/herself the directory from which the data has to be pulled. 
In particular, I would like to give two options:

- [x] provide only one directory: the dataset is automatically split into training and validation #65 
- [x] explicitly provide two directories: one directory is used to extract the training data, the other one is used to extract the validation data #101

### PR DESCRIPTION
**Pull request by pzhanggit**
Also rewrites `RawDataLoader` and no longer requires explicit match to known datasets

@pzhanggit you can close issues automatically with this:
Fixes #34

### CODE REVIEW

File: data_utils/raw_dataset_loader.py [hunk 1]
```diff
@@ -68,13 +68,25 @@ def load_raw_data(self, dataset_path: str, config):
         ) = self.__normalize_dataset(dataset=dataset)
 
         serial_data_name = config["name"]
-        serial_data_path = (
-            os.environ["SERIALIZED_DATA_PATH"]
-            + "/serialized_dataset/"
-            + serial_data_name
-            + ".pkl"
-        )
+        if dataset_append == "total":
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:71]: As we discussed, "combined"

File: data_utils/raw_dataset_loader.py [hunk 2]
```diff
@@ -68,13 +68,25 @@ def load_raw_data(self, dataset_path: str, config):
...
+            serial_data_path = (
+                os.environ["SERIALIZED_DATA_PATH"]
+                + "/serialized_dataset/"
+                + serial_data_name
+                + "_"
+                + dataset_append
+                + ".pkl"
+            )
 
+        # fixme; need to recalculate minimum and maximum values when train/test/validation sets provided separately
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:89]: We need this to merge, right?

File: data_utils/raw_dataset_loader.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ class RawDataLoader:
         Loads the raw files from specified path, performs the transformation to Data objects and normalization of values.
     """
 
-    def load_raw_data(self, dataset_path: str, config):
+    def load_raw_data(self, dataset_path: str, config, dataset_append=[]):
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:24]:
```suggestion
    def load_raw_data(self, dataset_path: str, config, dataset_type: str=""):
```

File: test_graphs.py [hunk 4]
```diff
@@ -7,14 +7,13 @@
...
 import torch
-import shutil
 
 torch.manual_seed(0)
 
 
 @pytest.mark.parametrize("model_type", ["GIN", "GAT", "MFC", "PNA"])
 @pytest.mark.parametrize("ci_input", ["ci.json", "ci_multihead.json"])
-def pytest_train_model(model_type, ci_input, overwrite_data=False):
+def pytest_train_model(model_type, ci_input, overwrite_data=True):
```
Comment (streeve) [inline test_graphs.py:16]: Rebase

File: utils/utils.py [hunk 5]
```diff
@@ -525,18 +535,24 @@ def combine_and_split_datasets(
 
 def load_data(dataset_option, config):
     transform_raw_data_to_serialized(config["Dataset"])
-    files_dir = (
-        f"{os.environ['SERIALIZED_DATA_PATH']}/serialized_dataset/{dataset_option}.pkl"
-    )
+    dataset_list = []
+    datasetname_list = []
+    for dataset_name, raw_data_path in config["Dataset"]["path"]["raw"].items():
```
Comment (streeve) [inline utils/utils.py:540]: `dataset_name` -> `dataset_type`

**Review by streeve — CHANGES_REQUESTED** [review]
Need to update `config.json`

File: data_utils/raw_dataset_loader.py (hunk 3 above)
Comment (allaffa) [reply data_utils/raw_dataset_loader.py:24]: Why's not use "None"?

File: data_utils/raw_dataset_loader.py (hunk 1 above)
Comment (allaffa) [reply data_utils/raw_dataset_loader.py:71]: "total" seems more clear to me.

Comment (allaffa) [reply data_utils/raw_dataset_loader.py:71]: or "full" or "complete" ?

Comment (streeve) [reply data_utils/raw_dataset_loader.py:71]: Let's just keep total then

**Comment by pzhanggit** [comment]
> Need to update `config.json`

Done in fa5774fa

File: data_utils/raw_dataset_loader.py (hunk 2 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py]: Done in 1599a85

File: data_utils/raw_dataset_loader.py (hunk 3 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py]: Done in f5053e816c483f06ec23529e692bf0bb7419109a

File: data_utils/raw_dataset_loader.py [hunk 6]
```diff
@@ -68,18 +68,117 @@ def load_raw_data(self, dataset_path: str, config):
...
 
         serial_data_name = config["name"]
-        serial_data_path = (
-            os.environ["SERIALIZED_DATA_PATH"]
-            + "/serialized_dataset/"
-            + serial_data_name
-            + ".pkl"
-        )
+        if dataset_type == "total":
+            serial_data_path = (
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:72]: `serial_data_name = config["name"] + ".pkl"`

File: data_utils/raw_dataset_loader.py [hunk 7]
```diff
@@ -68,18 +68,117 @@ def load_raw_data(self, dataset_path: str, config):
...
+                os.environ["SERIALIZED_DATA_PATH"]
+                + "/serialized_dataset/"
+                + serial_data_name
+                + ".pkl"
+            )
+        else:
+            # append for train; test; validation
+            serial_data_path = (
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:80]: `serial_data_name = config["name"] + "_" + dataset_type+ ".pkl"`
Then create `serial_data_path` outside

File: data_utils/raw_dataset_loader.py [hunk 8]
```diff
@@ -68,18 +68,117 @@ def load_raw_data(self, dataset_path: str, config):
...
+                + dataset_type
+                + ".pkl"
+            )
 
         with open(serial_data_path, "wb") as f:
             pickle.dump(minmax_node_feature, f)
             pickle.dump(minmax_graph_feature, f)
             pickle.dump(dataset_normalized, f)
 
+    def minmax_update(self, serial_data_name=None, data_types=None):
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:94]: As is, I think this function should be `SerializedDataLoader._normalize_dataset. Or we could to move the pickling outside `load_raw_data` and separately call `RawDataLoader._save_serialized_data` which would normalize, then dump the pickles

File: examples/ci.json [hunk 9]
```diff
@@ -5,7 +5,11 @@
     "Dataset": {
         "name": "unit_test",
         "format": "unit_test",
-        "path": "dataset/unit_test_singlehead",
+        "path": {
```
Comment (streeve) [inline examples/ci.json:8]: I like this for good unit test coverage

**Review by streeve — CHANGES_REQUESTED** [review]
Looks great, just want to rearrange a bit - let me know if you want me to make those changes

File: data_utils/raw_dataset_loader.py [hunk 10]
```diff
@@ -68,18 +68,117 @@ def load_raw_data(self, dataset_path: str, config):
...
-            + ".pkl"
-        )
+        if dataset_type == "total":
+            serial_data_path = (
+                os.environ["SERIALIZED_DATA_PATH"]
+                + "/serialized_dataset/"
+                + serial_data_name
+                + ".pkl"
+            )
+        else:
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:78]: We check the path key here. So it should be "total".

File: data_utils/raw_dataset_loader.py (hunk 8 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:94]: I am not sure I understand your comment. For multiple train/test/validate datasets, we need to gather the minimum and maximum values from all of them and then conduct normalization.

File: data_utils/raw_dataset_loader.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (streeve) [reply data_utils/raw_dataset_loader.py:78]: I forgot it would have already failed if wasn't total or test/train

File: data_utils/raw_dataset_loader.py [hunk 11]
```diff
@@ -21,7 +21,7 @@ class RawDataLoader:
         Loads the raw files from specified path, performs the transformation to Data objects and normalization of values.
     """
 
-    def load_raw_data(self, dataset_path: str, config):
+    def load_raw_data(self, dataset_path: str, config, dataset_type=None):
```
Comment (streeve) [inline data_utils/raw_dataset_loader.py:24]: str = "total" seems like a reasonable default

File: data_utils/raw_dataset_loader.py (hunk 7 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py]: updated in e64f94a12323

File: data_utils/raw_dataset_loader.py (hunk 11 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py]: updated

File: data_utils/raw_dataset_loader.py (hunk 8 above)
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py]: @streeve I made changes based on our discussion in 245ee5af5e5823b311d785ce2ef410f3167cba0c.

**Review by streeve — APPROVED** [review]
:heart: :heart: :heart:
